target_postgres/denest.py

```diff
@@ -0,0 +1,372 @@
+from copy import deepcopy
+
+from target_postgres import json_schema
+from target_postgres.singer_stream import (
+    SINGER_BATCHED_AT,
+    SINGER_LEVEL,
+    SINGER_PK,
+    SINGER_RECEIVED_AT,
+    SINGER_SEQUENCE,
+    SINGER_SOURCE_PK_PREFIX,
+    SINGER_TABLE_VERSION,
+    SINGER_VALUE
+)
+
+
+def to_table_batches(schema, key_properties, records):
+    """
+    Given a schema, and records, get all table schemas and records and prep them
+    in a `table_batch`.
+
+    :param schema: SingerStreamSchema
+    :param key_properties: [string, ...]
+    :param records: [{...}, ...]
+    :return: [{'streamed_schema': TABLE_SCHEMA(local),
+               'records': [{...}, ...]
+    """
+    table_schemas = _get_streamed_table_schemas(schema,
+                                                key_properties)
+
+    table_records = _get_streamed_table_records(key_properties,
+                                                records)
+    writeable_batches = []
+    for table_json_schema in table_schemas:
+        writeable_batches.append({'streamed_schema': table_json_schema,
+                                  'records': table_records.get(table_json_schema['path'], [])})
+
+    return writeable_batches
+
+
+def _get_streamed_table_schemas(schema, key_properties):
+    """
+    Given a `schema` and `key_properties` return the denested/flattened TABLE_SCHEMA of
+    the root table and each sub table.
+
+    :param schema: SingerStreamSchema
+    :param key_properties: [string, ...]
+    :return: [TABLE_SCHEMA(denested_streamed_schema_0), ...]
+    """
+    root_table_schema = json_schema.simplify(schema)
+
+    _add_singer_columns(root_table_schema, key_properties)
+
+    subtables = {}
+    key_prop_schemas = {}
+    for key in key_properties:
+        key_prop_schemas[key] = schema['properties'][key]
+    _denest_schema(tuple(), root_table_schema, key_prop_schemas, subtables)
+
+    ret = [_to_table_schema(tuple(), None, key_properties, root_table_schema['properties'])]
+    for path, schema in subtables.items():
+        ret.append(_to_table_schema(path, schema['level'], schema['key_properties'], schema['properties']))
+
+    return ret
+
+
+def _to_table_schema(path, level, keys, properties):
+    for key in keys:
+        if not (key,) in properties:
+            raise Exception('Unknown key "{}" found for table "{}". Known fields are: {}'.format(
+                key, path, properties
+            ))
+
+    return {'type': 'TABLE_SCHEMA',
+            'path': path,
+            'level': level,
+            'key_properties': keys,
+            'mappings': [],
+            'schema': {'type': 'object',
+                       'additionalProperties': False,
+                       'properties': properties}}
+
+
+def _add_singer_columns(schema, key_properties):
```

Comment:
Is this only used in the module? This seems like it's the function that handles this at each level